ssz_serialization/codec.nim

```diff
@@ -235,8 +237,20 @@ macro initSszUnionImpl(RecordType: type, input: openArray[byte]): untyped =
 func initSszUnion(T: type, input: openArray[byte]): T {.raisesssz.} =
   initSszUnionImpl(T, input)
 
+type HashListOperations = ref object
+  clearCaches: proc(idx: int) {.noSideEffect, raises: [].}
+  resetCache: proc() {.noSideEffect, raises: [].}
+  resizeHashes: proc() {.noSideEffect, raises: [].}
+
+func hashListClosures(hl: ptr auto): HashListOperations =
```

Comment:
The hashlist vs list code de facto ends up almost completely orthogonal - at this point, I think it's better to inline the hashlist invalidation in `elif val is HashList` and possibly extract any commonalities into helpers rather than resorting to these closures which make the code hard to read and introduce both additional memory allocations and indirect calls in what essentially is a non-dynamic code section..

I doubt this is a good/readable approach, but the closures can also trivially be removed by making the code generic, ie `type Context = object; readSszValue(input, val: var T, context = default(NoContext))` and then `readSszValue(data, myHashList.data, HashListContext(v: addr myHashList)` or something like that and `when context is HashListContext:` instead of `if`

Reply:
Yeah, agreed that would be a better approach.

Reply:
https://github.com/status-im/nim-ssz-serialization/pull/66/commits/30d073326e012fe99ece401dd7fe2021a32ad1b8
```
first (to fresh object) sDEI: (seconds: 0, nanosecond: 91918959)
first htr(seconds: 0, nanosecond: 622264520)
second (in-place) sDEI(seconds: 0, nanosecond: 21978290)
second htr (from updated object)(seconds: 0, nanosecond: 41531799)
third sDEI (in-place, back to initial state)(seconds: 0, nanosecond: 22590945)
third htr (of this in-place state revert)(seconds: 0, nanosecond: 42843366)
fourth sDEI (in-place, back to second state)(seconds: 0, nanosecond: 21427374)
fourth htr (back to second state)(seconds: 0, nanosecond: 41525163)
```
comparable to previous results, supposedly a bit faster, but probably noise from laptop-based measurement.

It's ultimately about 20 additional lines of code, so not that much, it's just repetitive.

There's some refactoring which could be done (some error detection, etc), but that's hindered by the `val`/`val.data` differences, so there's syntactic overhead there.

The generics approach might be worthwhile, yeah.

I'm not sure any of these are great, stylistically, but the issue is still that there are enough small, legitimate differences sprinkled throughout any approach is either duplicative or intricate. I'd go with a not-adding-fake-dynamic-behavior version at least, though, indeed.